File: src/main/java/rs/helpkit/JagexAppletViewer.java
package rs.helpkit;

import rs.helpkit.pref.RSPreferences;
import rs.helpkit.reflect.Classes;
import rs.helpkit.util.io.Internet;

import java.applet.Applet;
import java.io.File;
import java.io.IOException;
import java.lang.invoke.MethodHandle;
import java.lang.invoke.MethodHandles;
import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.net.JarURLConnection;
import java.net.URL;
import java.net.URLClassLoader;
import java.util.jar.JarFile;

/**
 * @author Tyler Sedlar
 * @since 3/20/2018
 */
public class JagexAppletViewer {

    private static final String VIEWER_URL = "http://oldschool.runescape.com/downloads/jagexappletviewer.jar";
    private static final String CONFIG = "http://oldschool%s.runescape.com/l=en/jav_config.ws";

    @SuppressWarnings("deprecation") // Applet is deprecated in Java9
    private static Field findAppletField(ClassLoader loader) throws ClassNotFoundException {
        Class<?> viewer = loader.loadClass("app.appletviewer");
        if (viewer == null) {
            throw new IllegalStateException("app.appletviewer not found");
        }
        Field applet = Classes.findField(viewer, field -> field.getType().equals(Applet.class));
        if (applet == null) {
            throw new IllegalStateException("Applet not found");
        }
        return applet;
    }

    @SuppressWarnings("deprecation") // Applet is deprecated in Java9
    private static Applet findRSApplet(ClassLoader loader) throws ClassNotFoundException, IllegalAccessException,
            InterruptedException {
        Field appletField = findAppletField(loader);

        Object applet;

        int tries = 0;

        while ((applet = appletField.get(null)) == null) { // check field value for non-null
            if (tries++ >= 10) { // something is clearly wrong is the applet hasn't loaded
                System.err.println("Unable to locate RS applet");
                System.exit(0);
            }
            Thread.sleep(1000);
        }

        return (Applet) applet;
    }

    public static void download(String targetFile) throws IOException {
        boolean outdated = true;

        File jar = new File(targetFile);

        if (jar.exists()) {
            int localHash = new JarFile(jar).getManifest().hashCode();
            String jarManifestURL = ("jar:" + VIEWER_URL + "!/META-INF/MANIFEST.MF");
            JarURLConnection remoteViewer = (JarURLConnection) new URL(jarManifestURL).openConnection();
            int remoteHash = remoteViewer.getManifest().hashCode();
            if (localHash == remoteHash) { // The local and remote manifest hashcode match, already up-to-date
                outdated = false;
            }
        }

        if (outdated) {
            // TODO: show download progress of jagexappletviewer.jar
            Internet.download(VIEWER_URL, jar.getAbsolutePath());
        } else {
            // TODO: show that jagexappletviewer.jar already up-to-date
        }
    }

    @SuppressWarnings("deprecation") // Applet is deprecated in Java9
    public static OSRSContainer run(String targetFile) throws Throwable {
        File jar = new File(targetFile);
        URLClassLoader loader = URLClassLoader.newInstance(new URL[]{jar.toURI().toURL()});

        Class<?> main = loader.loadClass("jagexappletviewer");
        Method exec = main.getDeclaredMethod("main", String[].class);

        System.setProperty("com.jagex.config", String.format(CONFIG, RSPreferences.getDefaultWorld()));
        System.setProperty("sun.java2d.nodraw", "true");

        MethodHandle mainHandle = MethodHandles.lookup().unreflect(exec);
        String[] viewerArgs = {"./"};
        mainHandle.invoke((Object[]) viewerArgs);

        Applet applet = findRSApplet(loader);

        return new OSRSContainer(applet);
    }
}

File: src/main/java/rs/helpkit/JagexAppletViewer.kt
package rs.helpkit

import rs.helpkit.api.util.Time
import rs.helpkit.internal.ReplacementTasks
import rs.helpkit.pref.RSPreferences
import rs.helpkit.reflect.Classes
import rs.helpkit.reflect.ObjectProxy
import rs.helpkit.util.io.Internet
import java.applet.Applet
import java.io.File
import java.io.IOException
import java.lang.invoke.MethodHandles
import java.lang.reflect.Field
import java.net.JarURLConnection
import java.net.URL
import java.net.URLClassLoader
import java.util.jar.JarFile

/**
 * @author Tyler Sedlar
 * @since 3/20/2018
 */
object JagexAppletViewer {

    private const val VIEWER_URL = "http://oldschool.runescape.com/downloads/jagexappletviewer.jar"
    private const val CONFIG = "http://oldschool%s.runescape.com/l=en/jav_config.ws"

    @Suppress("DEPRECATION") // Applet is deprecated in Java9
    @Throws(ClassNotFoundException::class)
    private fun findAppletField(loader: ClassLoader): Field {
        val viewer = loader.loadClass("app.appletviewer") ?: throw IllegalStateException("app.appletviewer not found")
        return Classes.findField(viewer) { field -> field.type == Applet::class.java }
                ?: throw IllegalStateException("Applet not found")
    }

    @Suppress("DEPRECATION") // Applet is deprecated in Java9
    private fun locateApplet(field: Field): Applet {
        for (i in 1..10) {
            val applet = field.get(null)
            if (applet != null) {
                return applet as Applet
            }
            Time.sleep(1000)
        }

        error("Unable to locate RS applet")
    }

    @Suppress("DEPRECATION") // Applet is deprecated in Java9
    @Throws(ClassNotFoundException::class, IllegalAccessException::class, InterruptedException::class)
    private fun findRSApplet(loader: ClassLoader): Applet {
        return locateApplet(findAppletField(loader))
    }

    @Throws(IOException::class)
    fun download(targetFile: String) {
        var outdated = true

        val jar = File(targetFile)

        if (jar.exists()) {
            val localHash = JarFile(jar).manifest.hashCode()
            val jarManifestURL = "jar:$VIEWER_URL!/META-INF/MANIFEST.MF"
            val remoteViewer = URL(jarManifestURL).openConnection() as JarURLConnection
            val remoteHash = remoteViewer.manifest.hashCode()
            if (localHash == remoteHash) { // The local and remote manifest hashcode match, already up-to-date
                outdated = false
            }
        }

        if (outdated) {
            // TODO: show download progress of jagexappletviewer.jar
            Internet.download(VIEWER_URL, jar.absolutePath)
        } else {
            // TODO: show that jagexappletviewer.jar already up-to-date
        }
    }

    @Throws(Throwable::class)
    fun run(targetFile: String): OSRSContainer? {
        val jar = File(targetFile)
        val loader = URLClassLoader.newInstance(arrayOf(jar.toURI().toURL()))

        val main = loader.loadClass("jagexappletviewer")

        val exec = main.getDeclaredMethod("main", Array<String>::class.java)

        System.setProperty("com.jagex.config", String.format(CONFIG, RSPreferences.defaultWorld))
        System.setProperty("sun.java2d.nodraw", "true")

        val mainHandle = MethodHandles.lookup().unreflect(exec)
        val viewerArgs = arrayOf("./")
        mainHandle.invoke(*viewerArgs as Array<*>)

        val applet = findRSApplet(loader)

        return OSRSContainer(applet)
    }
}
